File: holo-client/src/main/java/com/alibaba/hologres/client/Scan.java
/*
 * Copyright (c) 2020. Alibaba Group Holding Limited
 */

package com.alibaba.hologres.client;

import org.postgresql.model.TableSchema;

import java.security.InvalidParameterException;
import java.util.ArrayList;
import java.util.BitSet;
import java.util.List;

/**
 * Class to represent a Select operation (similar to select a,b,c from xxx where a=? and b=?).
 */
public class Scan {

	private final TableSchema schema;
	private List<Filter> filterList;
	private BitSet selectedColumns = null;
	private final int fetchSize;
	private final int timeout;
	private final SortKeys sortKeys;
	public Scan(TableSchema schema, List<Filter> filterList, BitSet selectedColumns, int fetchSize, int timeout, SortKeys sortKeys) {
		this.schema = schema;
		this.filterList = filterList;
		this.selectedColumns = selectedColumns;
		this.fetchSize = fetchSize;
		this.timeout = timeout;
		this.sortKeys = sortKeys;
	}

	public TableSchema getSchema() {
		return schema;
	}

	public List<Filter> getFilterList() {
		return filterList;
	}

	public BitSet getSelectedColumns() {
		return selectedColumns;
	}

	public int getFetchSize() {
		return fetchSize;
	}

	public int getTimeout() {
		return timeout;
	}

	public SortKeys getSortKeys() {
		return sortKeys;
	}

	public static Builder newBuilder(TableSchema schema) {
		return new Builder(schema);
	}

	/**
	 * Builder of Get request.
	 */
	public static class Builder {

		private final TableSchema schema;
		private List<Filter> filterList;
		private BitSet selectedColumns = null;
		private int fetchSize = -1;
		private int timeout = -1;
		private SortKeys sortKeys = SortKeys.PRIMARY_KEY;

		public Builder(TableSchema schema) {
			this.schema = schema;
		}

		public Builder addEqualFilter(String columnName, Object value) {
			Integer index = schema.getColumnIndex(columnName);
			if (index == null) {
				throw new InvalidParameterException(String.format("Table %s does not have column %s.", schema.getTableName(), columnName));
			}
			if (value == null) {
				throw new InvalidParameterException(String.format("Scan can not set null value for addEqualFilter. table %s ,column %s.", schema.getTableName(), columnName));
			}
			if (filterList == null) {
				filterList = new ArrayList<>();
			}
			filterList.add(new EqualsFilter(index, value));
			return this;
		}

		/**
		 * columnName >= start and columnName < end.
		 *
		 * @param columnName columnName
		 * @param start      nullable
		 * @param end        nullable
		 * @return this
		 */
		public Builder addRangeFilter(String columnName, Object start, Object end) {
			Integer index = schema.getColumnIndex(columnName);
			if (index == null) {
				throw new InvalidParameterException(String.format("Table %s does not have column %s.", schema.getTableName(), columnName));
			}
			if (start == null && end == null) {
				throw new InvalidParameterException(String.format("Scan can not set start and end both null. table %s ,column %s.", schema.getTableName(), columnName));
			}
			if (filterList == null) {
				filterList = new ArrayList<>();
			}
			filterList.add(new RangeFilter(index, start, end));
			return this;
		}

		public Builder withSelectedColumns(String[] columns) {
			for (String columnName : columns) {
				Integer index = schema.getColumnIndex(columnName);
				if (index == null) {
					throw new InvalidParameterException(String.format("Table %s does not have column %s.", schema.getTableName(), columnName));
				}
			}
			for (String columnName : columns) {
				this.withSelectedColumn(columnName);
			}
			return this;
		}

		public Builder withSelectedColumn(String columnName) {
			Integer index = schema.getColumnIndex(columnName);
			if (index == null) {
				throw new InvalidParameterException(String.format("Table %s does not have column %s.", schema.getTableName(), columnName));
			}
			if (selectedColumns == null) {
				selectedColumns = new BitSet(schema.getColumnSchema().length);
			}
			selectedColumns.set(index);
			return this;
		}

		public Builder setFetchSize(int fetchSize) {
			if (fetchSize <= 0) {
				throw new InvalidParameterException(String.format("FetchSize must > 0 . Table %s.", schema.getTableName()));
			}
			this.fetchSize = fetchSize;
			return this;
		}

		public Builder setTimeout(int timeout) {
			if (timeout <= 0) {
				throw new InvalidParameterException(String.format("Timeout must > 0 . Table %s.", schema.getTableName()));
			}
			this.timeout = timeout;
			return this;
		}

		public Builder setSortKeys(SortKeys sortKeys) {
			this.sortKeys = sortKeys;
			return this;
		}

		public Scan build() {
			return new Scan(schema, filterList, selectedColumns, fetchSize, timeout, sortKeys);
		}
	}
}

File: holo-client-py/hologres/scan.py
"""Scan API for prefix/filter-based queries.

Mirrors Java's Scan, EqualsFilter, RangeFilter, and SortKeys.

Usage::

    schema = client.get_table_schema("my_table")
    scan = (Scan.builder(schema)
            .add_equal_filter("id", 42)
            .set_sort_keys(SortKeys.NONE)
            .build())
    records = client.scan(scan)
"""

from __future__ import annotations

import enum
from dataclasses import dataclass, field
from typing import Any, List, Optional

from .table_schema import TableSchema


class SortKeys(enum.Enum):
    """Sort order for scan results."""

    PRIMARY_KEY = "PRIMARY_KEY"
    CLUSTERING_KEY = "CLUSTERING_KEY"
    NONE = "NONE"


@dataclass(frozen=True)
class EqualsFilter:
    """Equality filter: column = value."""

    column_index: int
    value: Any


@dataclass(frozen=True)
class RangeFilter:
    """Range filter: column >= start AND column < end (by default).

    Args:
        column_index: Index of the column in the schema.
        start: Start value (None for no lower bound).
        end: End value (None for no upper bound).
        start_inclusive: If True, use >= ; otherwise >.
        end_inclusive: If True, use <= ; otherwise <.
    """

    column_index: int
    start: Any = None
    end: Any = None
    start_inclusive: bool = True
    end_inclusive: bool = False


class Scan:
    """Represents a scan (prefix/filter query) operation.

    Use ``Scan.builder(schema)`` to create instances.
    """

    __slots__ = ("schema", "filters", "selected_columns", "sort_keys", "fetch_size")

    def __init__(
        self,
        schema: TableSchema,
        filters: List[EqualsFilter | RangeFilter],
        selected_columns: Optional[set[int]],
        sort_keys: SortKeys,
        fetch_size: int,
    ):
        self.schema = schema
        self.filters = filters
        self.selected_columns = selected_columns
        self.sort_keys = sort_keys
        self.fetch_size = fetch_size

    @classmethod
    def builder(cls, schema: TableSchema) -> ScanBuilder:
        """Create a builder for constructing a Scan."""
        return ScanBuilder(schema)

    def __repr__(self) -> str:
        return (
            f"Scan(table={self.schema.table_name}, "
            f"filters={len(self.filters)}, "
            f"sort_keys={self.sort_keys.name})"
        )


class ScanBuilder:
    """Builder for constructing Scan objects with validation."""

    __slots__ = (
        "_schema",
        "_filters",
        "_selected_columns",
        "_sort_keys",
        "_fetch_size",
    )

    def __init__(self, schema: TableSchema):
        self._schema = schema
        self._filters: List[EqualsFilter | RangeFilter] = []
        self._selected_columns: Optional[set[int]] = None
        self._sort_keys = SortKeys.PRIMARY_KEY
        self._fetch_size = 256

    def add_equal_filter(self, name: str, value: Any) -> ScanBuilder:
        """Add an equality filter (column = value)."""
        idx = self._schema.get_column_index(name)
        if idx is None:
            raise ValueError(f"Column {name!r} not found in schema")
        self._filters.append(EqualsFilter(column_index=idx, value=value))
        return self

    def add_range_filter(
        self,
        name: str,
        start: Any = None,
        end: Any = None,
        start_inclusive: bool = True,
        end_inclusive: bool = False,
    ) -> ScanBuilder:
        """Add a range filter."""
        idx = self._schema.get_column_index(name)
        if idx is None:
            raise ValueError(f"Column {name!r} not found in schema")
        self._filters.append(
            RangeFilter(
                column_index=idx,
                start=start,
                end=end,
                start_inclusive=start_inclusive,
                end_inclusive=end_inclusive,
            )
        )
        return self

    def with_selected_column(self, name: str) -> ScanBuilder:
        """Add a column to the projection."""
        idx = self._schema.get_column_index(name)
        if idx is None:
            raise ValueError(f"Column {name!r} not found in schema")
        if self._selected_columns is None:
            self._selected_columns = set()
        self._selected_columns.add(idx)
        return self

    def with_selected_columns(self, names: List[str]) -> ScanBuilder:
        """Add multiple columns to the projection."""
        for name in names:
            self.with_selected_column(name)
        return self

    def set_sort_keys(self, sort_keys: SortKeys) -> ScanBuilder:
        """Set the sort order for results."""
        self._sort_keys = sort_keys
        return self

    def set_fetch_size(self, fetch_size: int) -> ScanBuilder:
        """Set the fetch size for cursor-based iteration."""
        self._fetch_size = fetch_size
        return self

    def build(self) -> Scan:
        """Build and return the Scan."""
        return Scan(
            schema=self._schema,
            filters=list(self._filters),
            selected_columns=self._selected_columns,
            sort_keys=self._sort_keys,
            fetch_size=self._fetch_size,
        )
